File: construct_search.py
#!/usr/bin/env python
import sys
import time

t1 = t2 = lc = 0
d = dict()
with open(sys.argv[1],'r') as f:
  for line in f:
    start = time.time()
    d[line.rstrip()] = 'foo'
    t1 = t1 + (time.time() - start)
    lc += 1


miss = 0
with open(sys.argv[1],'r') as f:
  for line in f:
    key = line.rstrip()
    start = time.time()
    if key not in d and d[key] != 'foo':
      miss += 1
    t2 = t2 + (time.time() - start)


print "python\t%s\t%d\t%f\t%f\n" % (sys.argv[1],lc,t1,t2)

File: construct_search.R
args <- commandArgs(trailingOnly=TRUE)
if (length(args)!=2) stop("Need args: dataset Rname")
if (!file.exists(args[1])) stop("No file!")

e <- new.env(hash=TRUE, size=8*1024)

con <- file(args[1],open="r")
t1 <- 0.0
lc <- 0L
while(TRUE){
  line <- readLines(con,n=1)
  if (length(line) < 1) 
    break
  lc <- lc + 1L

  start <- proc.time()['elapsed']
  assign(line,"foo",envir=e);
  end <- proc.time()['elapsed']
  t1 <- t1 + end - start
}
close(con)
  

miss <- 0
t2 <- 0.0

con <- file(args[1],open="r")
while(TRUE){
  line <- readLines(con,n=1)
  if (length(line) < 1) 
    break

  start <- proc.time()['elapsed']
  m <- get(line,envir=e)
  end <- proc.time()['elapsed']
  t2 <- t2 + end - start

  # Miss should never miss!
  if (m!="foo") miss <<- miss + 1
}

cat(sprintf("%s\t%s\t%d\t%f\t%f\n",args[2],args[1],lc,t1,t2))
